design/one-pager-definition-of-done.md

```diff
@@ -0,0 +1,40 @@
+# Definition of done
+
+* Owner: Daniel Suskin (@suskin)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Introduction
+
+*Why does this document exist?*
+
+The goal of this document is to answer the questions:
+
+* As a developer, when should I consider a chunk of work to be complete?
+* As a reviewer, what should I be looking for to verify that a chunk of work is
+  complete?
+
+## Definition of work
+
+Work means a contribution to the project, typically code.
+
+## Definition of done
+
+A chunk of work is considered done when the following criteria have been met:
+
+* All code is written
+* Code is reviewed and approved
+* New behavior is documented in the documentation that will go onto the
+  crossplane website
+* All error conditions are reported into the log or as an event, as
+  appropriate
+* Code is tested, ideally in an integration test
```

Comment:
This reads like we prefer integration tests to unit tests, which I don't recall us agreeing on. Is it trying to suggest an integration test in addition to unit tests?

Reply:
Either way - is this actually something we expect today? We've only barely started establishing a pattern for integration testing in Crossplane. I would not personally know where to start if I were opening a new PR against, for example, one of the infrastructure stacks.

In general, I caution that [broken windows theory](https://en.wikipedia.org/wiki/Broken_windows_theory) (or a variant thereof) is [definitely a thing](https://github.com/crossplaneio/crossplane/pull/535#issuecomment-501458787) when it comes to checklists. If there's 1-2 things on the checklist that are skipped much of the time, then it becomes increasingly easy to ignore more and more of the checklist.  Therefore I would suggest:

* Keeping this list as short as possible.
* Breaking it up into _absolutely definitely required for every code change_ and _desired for every code change_.


We may consider looking for automation opportunities at some future time. One example - lint for PRs that touch code but that do not touch `docs/`.

Reply:
I agree with the point about broken windows here, and that it would be helpful to have a MUST and SHOULD section. I'll take a look into reworking the list to take that into account.

> This reads like we prefer integration tests to unit tests, which I don't recall us agreeing on. Is it trying to suggest an integration test in addition to unit tests?

This is something we should discuss. The reasoning behind this wording is that we don't know whether the changes really worked end-to-end until we write and execute an integration test. Having unit tests does increase our confidence, but doesn't necessarily guarantee that the behavior is correct from the perspective of a user scenario. What are our thoughts on whether we want to consider a chunk of work complete if we don't know whether the user scenario works?

A related discussion would be what we want to do to distinguish between the testing requirements for merging a changeset versus the testing requirements for promoting a changeset into a release. And how that relates to our definition of when a work item is done: if there **is** a distinction, is the work done when it is merged, or when it is released?

My personal preference is to consider work done when it is ready to be released.

Reply:
I have made some changes based on these ideas, but I believe the discussion about the place of integration tests and the difference between the merge gate and the release gate are discussions we will want to have before we settle on a definition of done.